zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -0,0 +1,181 @@
+
+use crate::{service::finalized_state::ZebraDb, DiskWriteBatch};
+
+use super::DiskFormatUpgrade;
+
+/// Implements [`DiskFormatUpgrade`] for adding Sapling and Orchard note commitment subtrees to database
+pub struct AddSubtrees;
+
+impl DiskFormatUpgrade for AddSubtrees {
+    fn run(
+        &self,
+        initial_tip_height: Height,
+        _network: Network,
+        upgrade_db: &ZebraDb,
+        should_cancel_format_change: Arc<AtomicBool>,
+    ) {
+        let Some((sapling_subtree_start_height, _)) = upgrade_db
+            .sapling_tree_by_height_range(Height(1)..)
+            .next() else { return };
+
+        let orchard_subtree_start_height = upgrade_db
+            .orchard_tree_by_height_range(Height(1)..)
+            .next()
+            .map(|(h, _)| h);
+
+        // Create an unbounded channel for reading note commitment trees
+        let (sapling_tree_tx, sapling_tree_rx) = mpsc::channel();
+
+        // Read sapling note commitment trees
+        let db = upgrade_db.clone();
+        let should_cancel_flag = should_cancel_format_change.clone();
+        let sapling_read_task = std::thread::spawn(move || {
+            let read_from = sapling_subtree_start_height.next();
+            for (height, tree) in db.sapling_tree_by_height_range(read_from..initial_tip_height) {
+                // Breaking from this loop and dropping the sapling_tree channel
+                // will cause the sapling compare and delete tasks to finish.
+                if should_cancel_flag.load(atomic::Ordering::Relaxed) {
+                    break;
+                }
+
+                if let Err(error) = sapling_tree_tx.send((height, tree)) {
+                    panic!("unexpected send error: {error:?}");
+                }
+            }
+        });
+
+        // Create an unbounded channel for sending sapling note commitment subtrees to be written
+        let (sapling_subtree_tx, sapling_subtree_rx) = mpsc::channel();
+
+        // Find sapling note commitment subtrees
+        let sapling_process_task = std::thread::spawn(move || {
+            let mut subtree_index = 0;
+            let mut start_height = sapling_subtree_start_height;
+            while let Ok((end_height, tree)) = sapling_tree_rx.recv() {
+                let tree_subtree_index = tree.subtree_index::<TRACKED_SUBTREE_HEIGHT>();
+
+                if let Some(subtree_root) = tree.subtree_root::<TRACKED_SUBTREE_HEIGHT>() {
+                    if let Err(error) = sapling_subtree_tx.send((
+                        subtree_index as u16,
+                        (start_height, end_height, subtree_root),
+                    )) {
+                        panic!("unexpected send error: {error:?}");
+                    }
+                }
+
+                if subtree_index < tree_subtree_index {
+                    start_height = end_height;
+                    subtree_index = tree_subtree_index;
+                }
+            }
```

Comment:
This is wrong.

Reply:
There's an "implementation details" section of ticket #6953 that might be helpful here, the details are currently behind a disclosure triangle.
<img width="544" alt="Screenshot 2023-08-21 at 14 24 30" src="https://github.com/ZcashFoundation/zebra/assets/8951843/75025db9-7a9a-424f-b9de-8d1abc22c9a8">